File: server/internal/mcpserver/projections_plaid.go
package mcpserver

import (
	"tallyo/internal/graph/model"
)

type leanPlaidCredentialRef struct {
	ID          int32   `json:"id"`
	ClientID    string  `json:"clientId"`
	Environment string  `json:"environment"`
	Label       *string `json:"label,omitempty"`
}

type leanPlaidCredential struct {
	leanPlaidCredentialRef
	ItemCount int32 `json:"itemCount"`
}

func mapPlaidCredentialRef(c *model.PlaidCredential) leanPlaidCredentialRef {
	return leanPlaidCredentialRef{
		ID:          c.ID,
		ClientID:    c.ClientID,
		Environment: string(c.Environment),
		Label:       c.Label,
	}
}

func mapPlaidCredential(c *model.PlaidCredential) leanPlaidCredential {
	return leanPlaidCredential{leanPlaidCredentialRef: mapPlaidCredentialRef(c), ItemCount: c.ItemCount}
}

func mapPlaidCredentialList(list *model.PlaidCredentialList) *leanList[leanPlaidCredential] {
	return mapList(list.Items, mapPlaidCredential)
}

type leanPlaidItem struct {
	ID              string                  `json:"id"`
	Credential      *leanPlaidCredentialRef `json:"credential,omitempty"`
	InstitutionID   *string                 `json:"institutionId,omitempty"`
	Accounts        []leanAccount           `json:"accounts"`
	HealthState     string                  `json:"healthState"`
	HealthErrorCode *string                 `json:"healthErrorCode,omitempty"`
	IsActive        bool                    `json:"isActive"`
}

func mapPlaidItem(item *model.PlaidItem) leanPlaidItem {
	lean := leanPlaidItem{
		ID:              item.ID.EncodedString(),
		InstitutionID:   item.InstitutionID,
		Accounts:        mapAccounts(item.Accounts),
		HealthState:     string(item.HealthState),
		HealthErrorCode: item.HealthErrorCode,
		IsActive:        item.IsActive,
	}
	if item.Credential != nil {
		credential := mapPlaidCredentialRef(item.Credential)
		lean.Credential = &credential
	}
	return lean
}

func mapPlaidItemList(list *model.PlaidItemList) *leanList[leanPlaidItem] {
	return mapList(list.Items, mapPlaidItem)
}

File: server-rs/src/mcpserver/projections_plaid.rs
use serde::Serialize;

use super::projections::{LeanAccount, LeanList, encode, map_accounts, map_list};
use crate::{
    graph::HydratedPlaidItem,
    ids::GlobalIdType,
    schema::{PlaidCredential, PlaidEnvironment, PlaidItemHealthState},
};

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub(super) struct LeanPlaidCredentialRef {
    pub(super) id: i32,
    pub(super) client_id: String,
    pub(super) environment: PlaidEnvironment,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub(super) label: Option<String>,
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub(super) struct LeanPlaidCredential {
    #[serde(flatten)]
    pub(super) credential: LeanPlaidCredentialRef,
    pub(super) item_count: i32,
}

fn map_plaid_credential_ref(credential: PlaidCredential) -> LeanPlaidCredentialRef {
    LeanPlaidCredentialRef {
        id: credential.id,
        client_id: credential.client_id,
        environment: credential.environment,
        label: credential.label,
    }
}

pub(super) fn map_plaid_credential_list(credentials: Vec<PlaidCredential>) -> LeanList<LeanPlaidCredential> {
    map_list(credentials, |credential| LeanPlaidCredential {
        item_count: credential.item_count,
        credential: map_plaid_credential_ref(credential),
    })
}

#[derive(Debug, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub(super) struct LeanPlaidItem {
    pub(super) id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub(super) credential: Option<LeanPlaidCredentialRef>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub(super) institution_id: Option<String>,
    pub(super) accounts: Vec<LeanAccount>,
    pub(super) health_state: PlaidItemHealthState,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub(super) health_error_code: Option<String>,
    pub(super) is_active: bool,
}

pub(super) fn map_plaid_item_list(items: Vec<HydratedPlaidItem>) -> LeanList<LeanPlaidItem> {
    map_list(items, |hydrated| {
        let HydratedPlaidItem {
            item,
            credential,
            accounts,
        } = hydrated;
        LeanPlaidItem {
            id: encode(GlobalIdType::PlaidItem, item.id),
            credential: credential.map(map_plaid_credential_ref),
            institution_id: item.institution_id,
            accounts: map_accounts(accounts),
            health_state: item.health_state,
            health_error_code: item.health_error_code,
            is_active: item.is_active,
        }
    })
}
